code/modules/paperwork/photocopier.dm

```diff
@@ -276,47 +176,321 @@
+		copyitem.forceMove(loc) //fuckery detected! get off my photocopier... shitbird!
+
+	copying = FALSE
+
+/obj/machinery/photocopier/proc/scan_document() //scan a document into a file
+	if(!cancopy())
+		return
+	if(length(saved_documents) >= max_saved_documents)
+		to_chat(usr, "<span class='warning'>\The [copyitem] can't be scanned because the max file limit has been reached. Please delete a file to make room.</span>")
+		return
+	copying = TRUE
+	var/obj/item/O
+	//Instead of calling copy() we jump ahead and use the procs that do the heavy lifting to avoid using toner since we're only scanning
+	if(istype(copyitem, /obj/item/paper))
+		O = papercopy(copyitem, scanning = TRUE)
+	else if(istype(copyitem, /obj/item/photo))
+		O = photocopy(copyitem, scanning = TRUE)
+	else if(istype(copyitem, /obj/item/paper_bundle))
+		O = bundlecopy(copyitem, scanning = TRUE, use_toner = FALSE)
+	else if(copymob && copymob.loc == loc)
+		O = copyass(scanning = TRUE)
+	else
+		to_chat(usr, "<span class='warning'>\The [copyitem] can't be scanned by \the [src].</span>")
+		copying = FALSE
+		return
+	use_power(active_power_usage)
+	sleep(PHOTOCOPIER_DELAY)
+	saved_documents.Add(O)
+	copying = FALSE
+	playsound(loc, 'sound/machines/ping.ogg', 50, 0)
+	atom_say("Document succesfully scanned!")
+
+/obj/machinery/photocopier/proc/delete_file(uid)
+	var/document = locateUID(uid)
+	if(document in saved_documents)
+		saved_documents.Remove(document)
+		qdel(document)
+
+/obj/machinery/photocopier/proc/file_copy(uid)
+	var/document = locateUID(uid)
+	if(document in saved_documents)
+		copy(document, scancopy = TRUE)
+
+
+/obj/machinery/photocopier/ui_interact(mob/user, ui_key = "main", datum/tgui/ui = null, force_open = FALSE, datum/tgui/master_ui = null, datum/ui_state/state = GLOB.default_state)
+	ui = SStgui.try_update_ui(user, src, ui_key, ui, force_open)
+	if(!ui)
+		ui = new(user, src, ui_key, "Photocopier", name, 402, 368, master_ui, state)
+		ui.open()
+
+/obj/machinery/photocopier/ui_data(mob/user)
+	var/list/data = list()
+	data["copynumber"] = copies
+	data["toner"] = toner
+	data["copyitem"] = (copyitem ? copyitem.name : null)
+	data["folder"] = (folder ? folder.name : null)
+	data["mob"] = (copymob ? copymob.name : null)
+	data["files"] = list()
+	for(var/obj/item/O in saved_documents)
+		var/list/document_data = list(
+			name = O.name,
+			uid = O.UID()
+		)
+		data["files"] += list(document_data)
+	return data
+
+/obj/machinery/photocopier/ui_act(action, list/params)
+	if(..())
+		return
+	add_fingerprint(usr)
+	switch(action)
+		if("copy")
+			copy(copyitem)
+		if("removedocument")
+			remove_document()
+		if("removefolder")
+			remove_folder()
+		if("add")
+			if(copies < maxcopies)
+				copies++
+		if("minus")
+			if(copies > 0)
+				copies--
+		if("scandocument")
+			scan_document()
+		if("filecopy")
+			file_copy(params["uid"])
+		if("deletefile")
+			delete_file(params["uid"])
+	update_icon()
```

Comment:
```suggestion
			delete_file(params["uid"])
	update_icon()
	return TRUE
```

Returning `TRUE` forces an UI update, which helps make things more snappy and responsive (especially when increasing/decreasing number of pages).
I typically